version/cmd/version/version.go

```diff
@@ -56,34 +40,54 @@ func main() {
 		flag.PrintDefaults()
 		os.Exit(1)
 	}
-	result, err := fn(tag)
+	result, err := fn(tag, bt)
 	if err != nil {
 		log.Fatal(err)
 	}
 	fmt.Println(result)
 }
 
-func getTag() (string, error) {
+// buildType getter func that matches the signature of other version funcs
+func buildType(_ string, bt version.BuildType) (string, error) {
+	return string(bt), nil
+}
+
+// implements the version.BuildEnv interface using the real build environment
+type BuildEnv struct{}
+
+// Returns true if we are building from a CI (rather than a local or dev build)
+func (b *BuildEnv) IsCI() bool {
+	// Travis, AppVeyor, CircleCI, and many others all set CI=true in their
+	// environment by default.
+	return os.Getenv("CI") == "true"
+}
+
+// Returns true if this is a nightly release by checking whether the current
+// HEAD is one or more commits ahead of the latest tag.
+func (b *BuildEnv) IsNightly() bool {
 	cmd := exec.Command("git", "describe", "--exact-match", "--tags", "HEAD")
-	out, err := cmd.Output()
-	if err != nil {
-		if _, ok := err.(*exec.ExitError); !ok {
```

Comment:
The reason this distinction exists is to separate non-zero exit status from all other errors that can occur while executing a command. (OOM for instance.)

In these circumstances, the error should be propagated up to the caller.

Reply:
The modified version of this function still distinguishes between exit error and other errors as you describe. It however does not propagate errors to the caller, which will be fixed shortly.